extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+   * @private
+   */
+  maybeRenderAmpAd_(valid, bytes) {
+    if (!valid) {
+      return Promise.resolve(false);
+    }
+    // Timer id will be set if we have entered layoutCallback at which point
+    // 3p throttling count was incremented.  We want to "release" the throttle
+    // immediately since we now know we are not a 3p ad.
+    if (this.timerId_) {
+      decrementLoadingAds(this.timerId_, this.getWin());
+    }
+    return xhrFor(this.getWin()).utf8FromArrayBuffer(bytes).then(creative => {
+      // Find the json blob located at the end of the body and parse it.
+      const creativeMetaData = this.getAmpAdMetadata_(creative);
+      if (!creativeMetaData || !this.supportsShadowDom()) {
+        // Shadow DOM is not supported or could not find appropriate markers
+        // within the creative therefore load within cross domain iframe.
+        // Iframe is created immediately (as opposed to waiting for
+        // layoutCallback) as the the creative has been verified as AMP and
+        // will run efficiently.
+        this.renderViaIframe_();
+        this.rendered_ = true;
+        return Promise.resolve(true);
+      } else {
+        try {
+          // Do extraction processing on CSS and body before creating the
+          // shadow root so that if they error out, we don't actually edit
+          // the doc.
+          const cssBlock = this.formatCSSBlock_(creative, creativeMetaData);
+          const bodyBlock = this.formatBody_(creative, creativeMetaData);
+          // Copy fonts to host document head.
+          this.relocateFonts_(creativeMetaData);
+          // Add extensions to head.
+          this.addCreativeExtensions_(creativeMetaData);
+          // Finally, add body and re-formatted CSS styling to the shadow root.
+          const shadowRoot =
+              this.element.shadowRoot || this.element.createShadowRoot();
+          shadowRoot./*OK*/innerHTML += (cssBlock + bodyBlock);
+          this.rendered_ = true;
+          this.onAmpCreativeShadowDomRender();
+          return Promise.resolve(true);
+        } catch (e) {
+          // If we fail on any of the steps of Shadow DOM construction, just
+          // render in iframe.
+          // TODO: report!
+          return Promise.resolve(false);
+        }
+      }
+    });
+  }
+
+  /**
+   * Creates iframe whose src matches that of the ad URL.  The response should
+   * have been cached causing the browser to render without callout.  However,
+   * it is possible for cache miss to occur which can be detected server-side
+   * by missing ORIGIN header.
+   * @param {boolean=} opt_isNonAmpCreative whether creative within iframe
+   *    is AMP creative (if not, intersection observer allows sending info into
+   *    nested frames).
+   * @private
+   */
+  renderViaIframe_(opt_isNonAmpCreative) {
+    if (!this.adUrl_) {
+      // Error should not occur.
+      return;
+    }
+    const iframe = this.element.ownerDocument.createElement('iframe');
```

Comment:
Ok, I've looked through that code in more detail.  Almost all of it seems to be to support the [AMP ad network API](https://github.com/ampproject/amphtml/blob/master/ads/README.md) (e.g., `no-content` and `entity-id` messages; maybe `embed-size`, though that's not documented in the API).  But the object inside this iframe won't be a 3p ad network implementation -- it will be an (arbitrary) HTML creative.  It won't know that it's being embedded in an AMP page iframe and won't be prepared to send / respond to such events, I don't think.

The visibility handling looks useful, but only if the `render-start` message is sent by the AMP runtime and not expected from the ad creative.  But I assume from the way this is written that it's expected to be coming from the iframe content?

Summary: I'm not sure how much, if any, of the code you outline is relevant in this case.